libr/io/io_bank.c

```diff
@@ -0,0 +1,669 @@
+	memset (buf, io->Oxff, len);
+	RIOSubMap *sm = node ? (RIOSubMap *)node->data : NULL;
+	while (sm && r_io_submap_overlap ((&fake_sm), sm)) {
+		RIOMap *map = r_io_map_get_by_ref (io, &sm->mapref);
+		if (!map) {
+			// mapref doesn't belong to map
+			return false;
+		}
+		if (!(map->perm & R_PERM_R)) {
+			continue;
+		}
+		const ut64 buf_off = addr - R_MAX (addr, r_io_submap_from (sm));
+		const int read_len = R_MIN (r_io_submap_to ((&fake_sm)),
+					     r_io_submap_to (sm)) - buf_off + 1;
+		const ut64 paddr = addr + buf_off - r_io_map_from (map) + map->delta;
+		r_io_fd_read_at (io, map->fd, paddr, &buf[buf_off], read_len);
+		// check return value here?
+		node = r_rbtree_cont_node_next (node);
+		sm = node ? (RIOSubMap *)node->data : NULL;
+	}
+	return true;
+}
+
+R_API bool r_io_bank_write_at(RIO *io, const ut32 bankid, ut64 addr, ut8 *buf, int len) {
+	RIOBank *bank = r_io_bank_get (io, bankid);
+	r_return_val_if_fail (io && bank, false);
+	RIOSubMap fake_sm;
+	memset (&fake_sm, 0x00, sizeof(RIOSubMap));
+	fake_sm.itv.addr = addr;
+	fake_sm.itv.size = len;
+	// TODO: handle overflow
+	RContRBNode *node = _find_entry_submap_node (bank, &fake_sm);
+	RIOSubMap *sm = node ? (RIOSubMap *)node->data : NULL;
+	while (sm && r_io_submap_overlap ((&fake_sm), sm)) {
+		RIOMap *map = r_io_map_get_by_ref (io, &sm->mapref);
+		if (!map) {
+			// mapref doesn't belong to map
+			return false;
+		}
+		if (!(map->perm & R_PERM_W)) {
+			continue;
+		}
+		const ut64 buf_off = addr - R_MAX (addr, r_io_submap_from (sm));
+		const int read_len = R_MIN (r_io_submap_to ((&fake_sm)),
+					     r_io_submap_to (sm)) - buf_off + 1;
+		const ut64 paddr = addr + buf_off - r_io_map_from (map) + map->delta;
+		r_io_fd_write_at (io, map->fd, paddr, &buf[buf_off], read_len);
+		// check return value here?
+		node = r_rbtree_cont_node_next (node);
+		sm = node ? (RIOSubMap *)node->data : NULL;
+	}
+	return true;
+}
+
+R_API void r_io_bank_delete_map (RIO *io, const ut32 bankid, const ut32 mapid) {
+//no need to check for mapref here, since this is "just" deleting
+	RIOBank *bank = r_io_bank_get (io, bankid);
+	RIOMap *map = r_io_map_get (io, mapid);
+	r_return_if_fail (bank && map);
+	RListIter *iter;
+	RIOMapRef *mapref = NULL;
+	r_list_foreach_prev (bank->maprefs, iter, mapref) {
+		if (mapref->id == map->id) {
+			_delete_submaps_from_bank_tree (io, bank, iter, map);
+			r_list_delete (bank->maprefs, iter);
+		}
+	}
+	// map is not referenced by this bank; nothing to do
+}
+
+// merges nearby submaps, that have a map ref to the same map, and free unneeded tree nodes
+R_API void r_io_bank_drain (RIO *io, const ut32 bankid) {
+	RIOBank *bank = r_io_bank_get (io, bankid);
+	r_return_if_fail (bank);
+	RContRBNode *node = r_rbtree_cont_node_first (bank->submaps);
+	RContRBNode *next = NULL;
+	while (node) {
+		next = r_rbtree_cont_node_next (node);
+		if (next) {
+			RIOSubMap *bd, *sm;
+			bd = (RIOSubMap *)node->data;
```

Comment:
define and assign in the same line

Reply:
but then it doesn't read as bdsm

Reply:
Oh then nvm. Thats clearly more important